---
title: "Interactions_R"
author: "Gabriela Arciszewska"
date: "2023-07-03"
output: html_document
---

# 1. DIFFERENTIAL GENE EXPRESSION ----
 Use DESeq2 to estimate the differential gene expression

_________________________ -----
LIBRARIES AND ENVIRONMENT ----

```{r}
setwd("~/Desktop/Dissertation/Analysis")
```

```{r}
library(tidyverse)
library(magrittr)
library(DESeq2)
library(pheatmap)
library(patchwork)
library(biomaRt)
library(reshape2)
```

```{r}
ensembl = useEnsembl(biomart = "ensembl",
                     dataset = "hsapiens_gene_ensembl")
```

_________________________ -----
# 2. DATA PROCESSING ----
Import count matrix and metadata ----
change path for the files and metadata

```{r}
raw_data <- read_csv("raw_data/markdup.featurecounts.csv") %>%
  column_to_rownames(var = "Geneid") # turn the geneid column into rownames
```

```{r}
metadata <- read_csv("resources/targets.csv") %>%
  mutate(condition = paste(treatment, treatment_length, dev_stage, sep = "_")) %>%
  #change all columns to factors
  modify_if(is.character, as.factor) %>%
  column_to_rownames(var = "sample")
```

re-order levels to make sure they're in the right order

```{r}
metadata$treatment <- relevel(metadata$treatment, "control")
metadata$dev_stage <- relevel(metadata$dev_stage, "D70")
```

MAKE SURE ALL ROWNAMES AND COLNAMES MATCH IN ORDER
```{r}
all(rownames(metadata) %in% colnames(raw_data))
```

# 3. Create deseq object for analysis ----

#a) Compare all controls day 70 vs all controls day 100

1. filter all controls for day 70 and day 100
2. filter matching column names of raw counts 
3. quality check - MAKE SURE ALL ROWNAMES AND COLNAMES MATCH IN ORDER

```{r}
metadata_c_D70D100 <- metadata[metadata$treatment == "control", ]
matching_columns <- intersect(colnames(raw_data), rownames(metadata_c_D70D100))
raw_c_D70D100 <- raw_data[, matching_columns]

sample_names_in_order <- rownames(metadata_c_D70D100)
colnames(raw_c_D70D100) <- sample_names_in_order

all(rownames(metadata_c_D70D100) %in% colnames(raw_c_D70D100))
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = raw_c_D70D100,
                              colData = metadata_c_D70D100,
                              design = ~ dev_stage)

```


#Filter low count genes ----
#remove genes that don't have at least 10 counts in 1/3 of the replicates #if metadata has 12 obs, then change it to 4 (12/3 = 4)

```{r}
keep <- rowSums(counts(dds) >= 10) >= 4
dds <- dds[keep,]
```

# _________________________ -----
#DIFFERENTIAL GENE EXPRESSION ----

#DGE and results ----


```{r}
identifiers <- getBM(values = rownames(dds),
                     filters = "ensembl_gene_id",
                     attributes = c("ensembl_gene_id", "hgnc_symbol", 
                                    "description", "entrezgene_id",
                                    "gene_biotype"),
                     mart = ensembl,
                     uniqueRows = TRUE) %>% 
  distinct(ensembl_gene_id, .keep_all = TRUE)
```

```{r}
library("writexl")

write_xlsx(identifiers ,"~/Desktop/Dissertation/Analysis/output_updated/geneset_enrichment/controls_D70_vsD100/identifiers.xlsx")

```

```{r}
dge <- DESeq(dds)
res <- results(dge,
               contrast = c("dev_stage", "D70", "D100"), #D70 numerator, D100 denominator, so I compare the change D70:D100
               alpha = 0.05) 

summary(res)
```

```{r}
#9NAs for pvalue and padj
res <- as.data.frame(res)
#By default, results assigns a p-value of NA to genes containing count outliers, as identified using Cook's distance. See the cooksCutoff argument for control of this behavior. Cook's distances for each sample are accessible as a matrix "cooks" stored in the assays() list. This measure is useful for identifying rows where the observed counts might not fit to a Negative Binomial distribution.

# Create a dataframe with only "log2FoldChange" values of "res" dataframe
logFC <- as.data.frame(res$log2FoldChange, row.names = row.names(res))
colnames(logFC) <- c("log2FoldChange")

```


```{r}
#changed this function, otherwise was getting an error of "the value "length of the condition > 1" - not sure why

#when log2FC value is na, then FC is assigned NA value
#when logFC >= 0, thenFC =  2^logFC
#when logFC <0, then FC =  -(1 / (2 ^ logFC)
applyAntiLogFC <- function(logFC) {
  antilog.2 <- ifelse(is.na(logFC), NA,
                      ifelse(logFC >= 0, 2 ^ logFC, -(1 / (2 ^ logFC))))
  antilog.2
}

```

```{r}
library(dplyr)

# Create a new column 'FC' using mutate()
logFC <- mutate(logFC, FC = applyAntiLogFC(log2FoldChange)) #no NAs

#pull out rows with gene ids as a column 
logFC <- rownames_to_column(logFC, var = "ensembl_gene_id")
row.names(logFC) <- 1:nrow(logFC)

#print dataframe
print(logFC)

```

```{r}

res <- rownames_to_column(res, var = "ensembl_gene_id")
row.names(res) <- 1:nrow(res)
res

```

```{r}
#get the "FC" column into the "res" dataframe and reorder so it comes right after "log2FoldChange"
res$FC <- logFC$FC 
res <- res %>%
  relocate(FC, .before = lfcSE)

res
```

```{r}
#add "entrezgene_id" and "gene_biotype" information

res <- left_join(res, identifiers, by = "ensembl_gene_id")
#adding extra column with the information on the type of regulation - upregulated, downregulated, no differentially expressed. Based on the filters:

#differentialy expressed -> gene which has a padj \< 0.05, and log2FoldChange and padj not NA
#no differentially expressed otherwise
#upregulated when res$log2FoldChange > 0 and downregulated when res$log2FoldChange < 0 

res$type <- ifelse(!is.na(res$log2FoldChange) & res$padj < 0.05 & !is.na(res$padj),
                   ifelse(res$log2FoldChange < 0, "down-regulation", "up-regulation"),
                   "no differential expression")

res
```

```{r}
# Volcano plot on the whole set of data. p adj <0.05 
exp_status <- ggplot(data.frame(res), aes(x = log2FoldChange, y = -log10(padj), color = res$type)) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  ggtitle("Volcano plot of differentially expressed genes") +
  xlab("log2 fold change") +
  ylab("-log10 p-value") +
  labs(color = "Differential Expression Status")

exp_status
```

```{r}
#set a working directory to the plots file
setwd("~/Desktop/Dissertation/Analysis/plots/controls")

#save a plot to a file
ggsave("exp_status_controls.png", exp_status)
```

```{r}
setwd("~/Desktop/Dissertation/Analysis/plots/controls")
png("heatmap_controls.png")

heatmap_controls <- heatmap(assay(dds[1:500]), scale = "row", labRow = FALSE, cexCol = 0.6)

#The heatmap represents gene expression patterns of the first 500 genes in the dataset across different samples. It allows to visualize the relative expression levels of these genes, potentially highlighting patterns or clusters in the data. Scaling by rows helps in comparing the expression levels of genes within each sample.

dev.off()

```

#PCA 

```{r}
dds_vst <- vst(dds)

pcaData <- plotPCA(dds_vst, intgroup = c("dev_stage", "treatment_length"), returnData = TRUE )

percentVar <- round(100 * attr(pcaData, "percentVar"))

PCA_controls <- ggplot(pcaData, aes(PC1, PC2, color=dev_stage, shape=treatment_length)) +
  ggtitle("PCA plot of the control samples") +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()

PCA_controls

#In PCA, each principal component explains a certain proportion of the total variance in the data. The higher the percentage explained by a principal component, the more important it is in capturing the overall variation. In this case, PC1 explains 40%  of variance compared, and PC2 explains 33%.

# PC2 (y-axis) separates the developmental stages of your samples, with day 100 on the top and day 70 on the bottom. This indicates that PC2 captures some of the variation related to the developmental stage of your samples.

# Regarding PC1 (x-axis), you mentioned that different shapes are observed within different colors (representing developmental stages) on this axis. This suggests that PC1 captures some of the variation related to the treatment length.

#set a working directory to the output file
setwd("~/Desktop/Dissertation/Analysis/plots/controls")

#save a plot to a file
ggsave("PCA_controls.png", PCA_controls)

```

#MA PLOT 
```{r}
library(ggplot2)
library(scales) # needed for oob parameter
library(viridis)

res$significant <- ifelse(res$padj < 0.05, "Significant", NA)

MA_plot <- ggplot(res, aes(baseMean, log2FoldChange, colour=padj)) + geom_point(size=1) + scale_y_continuous(limits=c(-3, 3), oob=squish) + scale_x_log10() + geom_hline(yintercept = 0, colour="darkorchid4", size=0.5, linetype="longdash") + labs(x="mean of normalized counts", y="log fold change") + scale_colour_viridis(direction=-1, trans='sqrt') + theme_bw() + geom_density_2d(colour="black", size=1) + ggtitle(" MA plot of the control samples") 

#set a working directory to the output file
setwd("~/Desktop/Dissertation/Analysis/plots/controls")

#save a plot to a file
ggsave("MA_plot_controls.png", MA_plot)
```

```{r}

normalized_counts <- counts(dge, normalized = TRUE) %>% as.data.frame()

normalized_counts <- normalized_counts %>% rename_with(~paste0("norm_counts_", .), everything())

normalized_counts <- rownames_to_column(normalized_counts, var = "ensembl_gene_id")

raw_c_D70D100 <- rownames_to_column(raw_c_D70D100, var = "ensembl_gene_id")

all_counts <- left_join (normalized_counts, raw_c_D70D100, by = "ensembl_gene_id") 

rownames(all_counts) <- all_counts[,1] 

all_counts <- dplyr::select(all_counts, - "ensembl_gene_id")

all_counts <- all_counts %>%
  rename_with(~str_replace(., "^(?!norm_counts_)", "raw_counts_"))

all_counts <- rownames_to_column(all_counts, var = "ensembl_gene_id")

res <- dplyr::select(res, - type)

final_results <- left_join(res, all_counts,by= "ensembl_gene_id")

```

```{r}
setwd("~/Desktop/Dissertation/Analysis/output_updated/DESeq_results/")

write.csv(final_results, file = "results_controls.csv", row.names = FALSE)
save(final_results, file = "results_controls.RData")

```
